ibis/expr/selectors.py

```diff
@@ -0,0 +1,286 @@
+"""Convenient column selectors.
+
+## Rationale
+
+Column selectors are convenience functions for selecting columns that share some property.
+
+## Discussion
+
+For example, a common task is to be able to select all numeric columns for a
+subsequent computation.
+
+Without selectors this becomes quite verbose and tedious to write:
+
+```python
+>>> t.select([t[c] for c in t.columns if t[c].type().is_numeric()])
+```
+
+Compare that to the [`numeric`][ibis.expr.selectors.numeric] selector:
+
+```python
+>>> t.select(s.numeric())
+```
+
+When there are multiple properties to check it gets worse:
+
+```python
+>>> t.select(
+...     [
+...         t[c] for c in t.columns
+...         if t[c].type().is_numeric()
+...         if ("a" in c.get_name() or "cd" in c.get_name())
+...     ]
+... )
+```
+
+Using a composition of selectors this is much less tiresome:
+
+```python
+>>> t.select(s.numeric() & s.contains(("a", "cd")))
+```
+"""
+
+from __future__ import annotations
+
+import inspect
+import re
+from typing import Callable, Iterable, Sequence
+
+from public import public
+
+import ibis.expr.datatypes as dt
+import ibis.expr.types as ir
+from ibis import util
+
+
+class Selector:
```

Comment:
Could we subclass deferred here? I mean a deferred base to not provide all of the overloads. 

Mostly just asking, shouldn't block this PR.

Reply:
Maybe, but I didn't think much about what the semantics of the other overloads would be. It did cross my mind that maybe there's some shared code between this and `Deferred` but I didn't want to do a bunch of refactoring here.